base/compiler/optimize.jl

```diff
@@ -453,9 +457,262 @@ function finish(interp::AbstractInterpreter, opt::OptimizationState,
     return nothing
 end
 
+function visit_bb_phis!(callback, ir::IRCode, bb::Int)
+    stmts = ir.cfg.blocks[bb].stmts
+    for idx in stmts
+        stmt = ir[SSAValue(idx)][:inst]
+        if !isa(stmt, PhiNode)
+            if !is_valid_phiblock_stmt(stmt)
+                return
+            end
+        else
+            callback(idx, stmt)
+        end
+    end
+end
+
+function any_stmt_may_throw(ir::IRCode, bb)
+    for stmt in ir.cfg.blocks[bb].stmts
+        if (ir[SSAValue(stmt)][:flag] & IR_FLAG_NOTHROW) != 0
+            return true
+        end
+    end
+    return false
+end
+
+function conditional_successors_may_throw(lazypostdomtree::LazyPostDomtree, ir::IRCode, bb::Int)
+    visited = BitSet((bb,))
+    worklist = Int[bb]
+    postdomtree = get!(lazypostdomtree)
+    while !isempty(worklist)
+        thisbb = pop!(worklist)
+        for succ in ir.cfg.blocks[thisbb].succs
+            succ in visited && continue
+            push!(visited, succ)
+            postdominates(postdomtree, succ, thisbb) && continue
+            any_stmt_may_throw(ir, succ) && return true
+            push!(worklist, succ)
+        end
+    end
+    return false
+end
+
+function ipo_dataflow_analysis!(interp::AbstractInterpreter, ir::IRCode, result::InferenceResult)
+    inconsistent = BitSet()
+    inconsistent_bbs = BitSet()
+    tpdum = TwoPhaseDefUseMap(length(ir.stmts))
+    lazypostdomtree = LazyPostDomtree(ir)
+
+    all_effect_free = true # TODO refine using EscapeAnalysis
+    all_nothrow = true
+    all_retpaths_consistent = true
+    all_noub = true
+    any_conditional_ub = false
+    had_trycatch = false
+
+    scanner = BBScanner(ir)
+
+    effects = result.ipo_effects
+
+    cfg = nothing
+    domtree = nothing
+    function get_augmented_domtree()
+        if cfg !== nothing
+            return (cfg, domtree)
+        end
+        cfg = copy(ir.cfg)
+        # Add a virtual basic block to represent the exit
+        push!(cfg.blocks, BasicBlock(StmtRange(0:-1)))
+
+        for bb = 1:(length(cfg.blocks)-1)
+            terminator = ir[SSAValue(last(cfg.blocks[bb].stmts))][:inst]
+            if isa(terminator, ReturnNode) && isdefined(terminator, :val)
+                cfg_insert_edge!(cfg, bb, length(cfg.blocks))
+            end
+        end
+
+        domtree = construct_domtree(cfg.blocks)
+        return (cfg, domtree)
+    end
+
+    function is_getfield_boundscheck(inst::Instruction)
+        stmt = inst[:stmt]
+        is_known_call(stmt, getfield, ir) || return false
+        length(stmt.args) < 4 && return false
+        boundscheck = stmt.args[end]
+        argextype(boundscheck, ir) === Bool || return false
+        isa(boundscheck, SSAValue) || return false
+        return true
```

Comment:
Include `isexpr(ir[boundscheck][:stmt], :boundscheck)` here?

Reply:
This is meant to check for having a boundscheck argument rather than the boundscheck expr, I'll rename the function.